MarkBind/markbind
# Implement an api to ignore text in certain tags

**What is the purpose of this pull request? (put "X" next to an item, remove the rest)**

• [x] New feature

Groundwork for #984 

**What is the rationale for this request?**
To provide a convenient interface, for plugins / core markbind code to blacklist certain tags to ignore in parsers.

This should open the doors to custom components that can have all sorts of strange syntaxes inside them, without using markdown code blocks. e.g. `puml` #968 

**What changes did you make? (Give an overview)**
- Patch an existing _rule_ in markdown-it, exposing an interface that allows injecting tags to ignore before registering the patch. The patch allows these additional tags to be treated the same way `<script/pre/style>` tags are treated in markdown-it currently.
- Patch many of htmlparser2 state handlers and added some state variables to the tokenizer, exposing a similar interface.

**Provide some example code that this change will affect:**


```js
function injectIgnoreTags(tagsToIgnore) {
	Tokenizer.prototype.specialTagNames = Tokenizer.prototype.specialTagNames.concat(...tagsToIgnore)
}
```

**Is there anything you'd like reviewers to focus on?**
https://github.com/MarkBind/markbind/issues/984#issuecomment-586812974 - degree of commonmark compliance we should follow

**Testing instructions:**
- `npm run test` should work
- rebuilding the 2103 site should show no diffs, save for timestamps and log output

**Proposed commit message: (wrap lines at 72 characters)**
Implement an api to ignore content in special tags

The parsers MarkBind uses parses content in html tags as html
or markdown respectively.
This makes it difficult to add custom components that utilize
conflicting syntax.

Let’s add an interface to ignore content in certain special tags.
This interface is also exposed to plugins as the getSpecialTags option.
